### PR TITLE
Heroku deployment configuration

### PR DESCRIPTION
**Pull request by mvandenburgh**
This PR overhauls the current dandidav deployment process. Currently, we have Heroku configured in a "pull" model, i.e. Heroku constantly pulls our repo and reacts to new commits on `main` by doing a new deployment. This PR changes this to a "push" model, where we instead explicitly build and deploy the application on push to `main` via the Heroku CLI in a GitHub Action. Note, a side effect of this is we have complete control over the environment since we're running `cargo build` in the GitHub Action as opposed to on Heroku; thus, the commit hash issue is no longer a problem.

It contains two components:

- A `Procfile` for Heroku
	- This defines the command for the web dyno to run in order to start the server
- The GitHub Action that performs the deployment with the Heroku CLI whenever a PR with a "release" label is merged

Other things that need to be done before merging this PR:
- [x] ~~Merge the dandi-infrastructure PR removing the Rust buildpack (https://github.com/dandi/dandi-infrastructure/pull/170)~~
- [x] ~~Set the `HEROKU_API_KEY` and `HEROKU_EMAIL` secrets on the `dandi/dandidav` repo~~

### CODE REVIEW

File: .github/workflows/deploy.yml [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  CARGO_TERM_COLOR: always
+
+jobs:
+  deploy:
+    name: Deploy to Heroku
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0 # fetch history for all branches and tags
```
Comment (jwodder) [inline .github/workflows/deploy.yml:24]: This is unnecessary.

File: .github/workflows/deploy.yml [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+      - name: Compile the Rust appication
```
Comment (jwodder) [inline .github/workflows/deploy.yml:26]:
```suggestion
      - name: Compile the Rust application
```

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        run: cargo build --release
+
+      - name: Install Heroku CLI
+        run: curl https://cli-assets.heroku.com/install.sh | sh
+
+      - name: Install builds plugin
+        run: heroku plugins:install heroku-builds
+
+      - name: Create Heroku Build
+        run: heroku builds:create --app dandidav
```
Comment (jwodder) [inline .github/workflows/deploy.yml:36]: Is this deploying the entire repository, including the whole of the `target/` directory?  Can we not just deploy only the `target/release/dandidav` asset?

File: .github/workflows/deploy.yml [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+        env:
+          HEROKU_API_KEY: ${{ secrets.HEROKU_API_KEY  }}
+          HEROKU_EMAIL: ${{ secrets.HEROKU_EMAIL  }}
```
Comment (jwodder) [inline .github/workflows/deploy.yml:39]:
```suggestion
          HEROKU_API_KEY: ${{ secrets.HEROKU_API_KEY }}
          HEROKU_EMAIL: ${{ secrets.HEROKU_EMAIL }}
```

**Review by jwodder — CHANGES_REQUESTED** [review]
I would rather not use `auto` at all.  Instead, deployment should be triggered by any merge of a PR that does not have the "skip deployment" label (See [here](https://github.com/datalad/release-action/blob/20970e59c38f1a0bda80d4dfb4a0bbf82f9a2a14/populate-workflows.sh#L66) for a similar trigger definition).

**Comment by jwodder** [comment]
@mvandenburgh 

> Set the `HEROKU_API_KEY` and `HEROKU_EMAIL` secrets on the `dandi/dandidav` repo

Set them to what values?

**Comment by jwodder** [comment]
@mvandenburgh Also, please enable caching of dependencies etc. [like is done in `test.yml`](https://github.com/dandi/dandidav/blob/95b1ac29a105e07997a86afdc94aaee153c93c62/.github/workflows/test.yml#L56-L58).

File: .github/workflows/deploy.yml (hunk 3 above)
Comment (mvandenburgh) [reply .github/workflows/deploy.yml]: Unfortunately the `heroku builds` plugin doesn't allow specifying a single asset. And additionally, you have to include the `Procfile`. 

I didn't realize how many files got stored in the `target/` directory until I just ran the compilation locally, and I agree it's not ideal to push all those artifacts to Heroku. I tried getting around it by storing things in a temporary directory and pushing that - https://github.com/dandi/dandidav/pull/105/commits/6b8df2c06453999b1a2bca17796afb1d1452f0b1

**Comment by mvandenburgh** [comment]
> @mvandenburgh
> 
> > Set the `HEROKU_API_KEY` and `HEROKU_EMAIL` secrets on the `dandi/dandidav` repo
> 
> Set them to what values?

I thought we needed someone with admin permissions to set those, but it appears my credentials suffice. So this is all set now

**Comment by mvandenburgh** [comment]
@jwodder I believe I addressed all your feedback, PTAL and let me know if you have any further questions and/or change requests.

File: .github/workflows/deploy.yml [hunk 5]
```diff
@@ -0,0 +1,50 @@
+name: Deploy `dandidav` to Heroku
+
+on:
+  push:
+    branches:
+      - main
```
Comment (jwodder) [inline .github/workflows/deploy.yml:6]:
```suggestion
  pull_request_target:
    branches:
      - main
    types:
      - closed
```

This is needed for the `if:` below to work properly.

File: .github/workflows/deploy.yml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+jobs:
+  deploy:
+    name: Deploy to Heroku
+    runs-on: ubuntu-latest
+    if: "github.event.pull_request.merged == true && contains(github.event.pull_request.labels.*.name, 'release')"
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Activate cache
+        if: "!startsWith(github.head_ref, 'dependabot/')"
```
Comment (jwodder) [inline .github/workflows/deploy.yml:27]: This line is irrelevant to this workflow.

**Review by jwodder — CHANGES_REQUESTED** [review]

**Review by jwodder — APPROVED** [review]

**Pull request merged**